### PR TITLE
Refactor XmppProvider and component discovery

### PR DESCRIPTION
**Pull request by bgrozev**
- ref: Move XmppProvider to org.jitsi.jicofo.xmpp.
- ref: Merge AbstractXmppProvider into XmppProvider.
- ref: Merge XmppProviderImpl into XmppProvider.
- ref: Extract a Features enum, merge DiscoveryUtil into XmppProvider.
- ref: Move RegistrationListener to XmppProvider.Listener.
- ref: Move component discovery to XmppProvider.

### CODE REVIEW

File: jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt [hunk 1]
```diff
@@ -58,7 +145,51 @@ abstract class XmppProvider(val config: XmppConnectionConfig, parentLogger: Logg
         registrationListeners.remove(listener)
     }
 
-    protected open fun fireRegistrationStateChanged(registered: Boolean) {
+    /**
+     * Method tries to establish the connection to XMPP server and return
+     * <tt>false</tt> in case we have failed want to retry connection attempt.
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt:156]: Also, the semantics are backwards 🤦

File: jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt [hunk 2]
```diff
@@ -58,7 +145,51 @@ abstract class XmppProvider(val config: XmppConnectionConfig, parentLogger: Logg
...
+                    .addReconnectionListener(reconnectionListener)
+                if (config.password != null) {
+                    val login = config.username.toString()
+                    val pass = config.password
+                    val resource = config.resource
+                    xmppConnection.login(login, pass, resource)
+                }
+                false
+            } catch (e: java.lang.Exception) {
+                logger.error("Failed to connect/login: " + e.message, e)
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt:181]: I prefer to include both to have the message on the same line for easier grepping. I'll change to use string interpolation, I missed this one.

File: jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt [hunk 3]
```diff
@@ -18,30 +18,117 @@
...
+    private val connectionListener = object : ConnectionListener {
+        override fun authenticated(connection: XMPPConnection?, resumed: Boolean) { registered = true }
+
+        override fun connectionClosed() {
+            logger.info("XMPP connection closed")
+            registered = false
+        }
+
+        override fun connectionClosedOnError(e: Exception) {
+            logger.error("XMPP connection closed on error: ${e.message}", e)
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt:113]: Without including message it formats something like this:
```
XMPP connection closed on error
java.lang.Exception: MESSAGE
  at ...
```

File: jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt [hunk 4]
```diff
@@ -71,24 +202,115 @@ abstract class XmppProvider(val config: XmppConnectionConfig, parentLogger: Logg
...
+    return connection
+}
+
+private class Muc(val xmppProvider: XmppProvider) {
+    /** The active chat rooms mapped by their name. */
+    private val rooms: MutableMap<String, ChatRoomImpl> = HashMap()
+
+    @Throws(RoomExistsException::class)
+    fun createChatRoom(roomJid: EntityBareJid): ChatRoom {
+        val roomName = roomJid.toString().lowercase(Locale.getDefault())
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt:327]: Hmm I think that's how IDEA translated java to kotlin. I'll take another look to see if I can just use the `Jid` as the key.

File: jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt [hunk 5]
```diff
@@ -132,16 +147,9 @@ class XmppProvider(val config: XmppConnectionConfig, parentLogger: Logger) {
      * Registers the specified listener with this provider so that it would receive notifications on changes of its
      * state.
      */
-    fun addListener(listener: Listener) = synchronized(listeners) {
-        if (!listeners.contains(listener)) {
-            listeners.add(listener)
-        }
-    }
-
+    fun addListener(listener: Listener) = listeners.add(listener)
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/XmppProvider.kt:150]: I changed it to a CopyOnWriteArraySet

**Review by JonathanLennox — APPROVED** [review]

**Pull request merged**